pallets/payroll/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Payroll pallet
+//!
+//! ## Overview
+//!
+//! This pallet is part of DAOFi modules that provides payroll management
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+    pallet_prelude::*,
+    traits::{
+        tokens::fungibles::{Inspect, Mutate, Transfer},
+        UnixTime,
+    },
+    transactional,
+    weights::DispatchClass,
+    PalletId,
+};
+use frame_system::pallet_prelude::*;
+use scale_info::TypeInfo;
+use primitives::*;
+use sp_runtime::{
+    traits::{
+        AccountIdConversion,
+        Zero,
+    },
+    ArithmeticError, DispatchError
+};
+use sp_std::{prelude::*};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod weights;
+
+type AssetIdOf<T> =
+    <<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+type BalanceOf<T> =
+    <<T as Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+type AccountOf<T> = <T as frame_system::Config>::AccountId;
+
+// Struct for Payroll stream
+#[derive(Clone, Encode, Decode, PartialEq, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+#[codec(mel_bound())]
+pub struct Stream<T: Config> {
+    // Remaining Balance
+    remaining_balance: BalanceOf<T>,
+    // Deposit
+    deposit: BalanceOf<T>,
+    // Currency Id
+    currency_id: AssetIdOf<T>,
+    // Rate Per Second
+    rate_per_sec: BalanceOf<T>,
+    // Recipient
+    recipient: AccountOf<T>,
+    // Sender
+    sender: AccountOf<T>,
+    // Start Time
+    start_time: Timestamp,
+    // Stop Time
+    stop_time: Timestamp,
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    use weights::WeightInfo;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Assets for deposit/withdraw collateral assets to/from loans module
+        type Assets: Transfer<Self::AccountId, AssetId = CurrencyId, Balance = Balance>
+            + Inspect<Self::AccountId, AssetId = CurrencyId, Balance = Balance>
+            + Mutate<Self::AccountId, AssetId = CurrencyId, Balance = Balance>;
+            
+        /// The payroll module id, keep all collaterals of CDPs.
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+
+        /// Unix time
+        type UnixTime: UnixTime;
+
+        /// Weight information
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        StreamToOrigin,
```

Comment:
Can we change it to something like `NotAValidOrigin` or `NotAValidSender`

Reply:
`StreamToOrigin` error happens only on creating stream to sender, so I think `NotAValidOrigin` or `NotAValidSender` may be too generic.

Reply:
Here https://github.com/parallel-finance/parallel/blob/9d34d8acb3ca85214c71dcba2cb541d0ca95e086/pallets/payroll/src/lib.rs#L170 we are matching recipient with sender. `StreamToOrigin` looks bit confusing to me.

@GopherJ What do you think?

Reply:
Yes. This works